[Tooling] Make `pod trunk push` be `--synchronous`, to fix issue when releasing both pods together

AliSoftware opened the pull request:
## Why?

To prevent the Editor pod to fail to be pushed to trunk if the Aztec pod it depends on hasn't propagated through the CDN yet

Fixes the issue raised in this Slack thread: p1711026617605489-slack-CC7L49W13 cc @geriux 

## Testing

It's gonna be difficult to test the change on this repo without actually doing a new version of the Aztec and Editor pods.

The new ability of `publish_pod` to accept a `--synchronous` flag has already been tested as part of the Gravatar-iOS-SDK repo (which had a similar issue of co-dependant pods having to be released together), so I think we should be good though.

So I say we can probably trust that it will work as expected like it did for the Gravatar repo, and I'm counting on the next dev who will have to make a new version of these pods to let us know if it ends up still not working.

AliSoftware committed:
Make pod trunk push be --synchronous

To prevent the Editor pod to fail to be pushed to trunk if the Aztec pod it depends on hasn't propagated through the CDN yet

mokagio commented:
I noticed CI has been waiting for almost 12 hours.

Not surprising given it's configured to run on Xcode 13 😳 

<img width="239" alt="image" src="https://github.com/wordpress-mobile/AztecEditor-iOS/assets/1218433/d77f41c0-e9fb-47de-b7f5-7d1f88b28513">

Looking at upgrading now...

mokagio committed:
DRY `env` and `agents` in Buildkite pipeline

mokagio committed:
Address an access control warning

> Non-'@objc' instance method in extensions cannot be overridden;
> use 'public' instead

mokagio committed:
Centralize deployment target definition at projects level

mokagio committed:
Bump deployment target to iOS 12.0, minimum for Xcode 15.3

mokagio committed:
Use Xcode 15.3 in CI

mokagio committed:
Use Ruby 3.2.2 and update RuboCop to compatible version

mokagio committed:
Bump iOS and Simulator versions for CI

mokagio committed:
Allow warnings when validating podspec

Ideally, this should be removed once the warnings have been addressed

mokagio committed:
Update Fastlane in the hope it helps with stuck tests in CI

mokagio committed:
Don't waste time cleaning up Simulators in CI

AliSoftware committed:
Merge branch 'develop' into tooling/fix-sync-pod-publication

mokagio committed:
Merge remote-tracking branch 'origin/develop' into mokagio/xcode-15.3

AliSoftware commented:
It seems like the test failures are due to the "Allow Paste" System Prompt—which now appears on the simulator in more recent iOS versions—during the copy/paste-related tests.

@geriux @mokagio can I let you look at those failures? There's probably a way in XCTest to catch those during a UITest (maybe [`addUIInterruptionMonitor `](https://developer.apple.com/documentation/xctest/xctestcase/1496273-adduiinterruptionmonitor)? Other?)
I gave it a quick try on my end but this quickly turned into a rabbit hole and I didn't have the bandwidth to go further on my end, so I'll let this to you 🙇 

---

PS: While taking a quick look at the workspace I noticed that the build settings are still set to `DEPLOYMENT_TARGET=11.0` (and the `podspec` as well), and if we update the project to e.g. iOS13, there are a couple of depreciations warnings and changes that would be needed in the code to get rid of those warnings (which would otherwise prevent the pod to be released) + make the lib ready for later updates.

I also noticed that all the methods around manipulating and converting between `NSRange`/`UTF16NSRange`/`Range<String.UTF16View>` seems from a time where the model for the `String` type in Swift was not as nice as today's. Since then, `Swift.Foundation` have gotten a lot of built-in methods for that (like `NSRange(swiftrange, in: swiftstring)` and the likes) which would be more resilient (including handling edge cases like `NSNotFound` and how it might be represented differently in 32 vs 64 bits platforms, ranges with negative locations and/or length, etc) than the manual code written in Aztec's extensions, and switching to them would make the code more bug-proof, especially on such a tricky topic of manipulating Unicode representations of Strings.

mokagio commented:
> PS: While taking a quick look at the workspace I noticed that the build settings are still set to DEPLOYMENT_TARGET=11.0 (and the podspec as well), and if we update the project to e.g. iOS13, there are a couple of depreciations warnings and changes that would be needed in the code to get rid of those warnings (which would otherwise prevent the pod to be released) + make the lib ready for later updates.

2f50d83de2c2d6d2c9f6ed7f9ec73eb0fa9d3e4e bumps the target to iOS 12, which is the minimum supported by Xcode 15. The warnings remain, I added `--allow-warnings` to the `pod` commands as a workaround.

mokagio commented:
> It seems like the test failures are due to the "Allow Paste" System Prompt—which now appears on the simulator in more recent iOS versions—during the copy/paste-related tests.

🤔 

@jkmassel addressed this in #1377 by making the pasteboard injectable during unit testing. When I merged the latest `develop` into #1392, the lockup related to prompt disappeared.

As far as I can tell, the remaining tests failures at the custom `TextView` level are due to something misbehaving during the HTML parsing. If I had time to investigate, I would pursue the `String` API route you suggested:

> I also noticed that all the methods around manipulating and converting between NSRange/UTF16NSRange/Range<String.UTF16View> seems from a time where the model for the String type in Swift was not as nice as today's

mokagio committed:
Merge remote-tracking branch 'origin/tooling/fix-sync-pod-publication' into mokagio/xcode-15.3

AliSoftware commented:
>  The warnings remain, I added `--allow-warnings` to the `pod` commands as a workaround.
I don't see that `--allow-warnings` added to the pod commands in `.buildkite/` yet, so I'm guessing you either wrote that you'd do it but forgot, or forgot to push the change 😄

AliSoftware committed:
Fix NSSecureCoding implementations

@mokagio This should fix the `testPaste{Image,Video}WithoutFormatting` test failures.
 - I have searched the code for `decodeObject(forKey` occurrences and confirmed there wasn't any left
 - But I have not triple-checked that *every* custom subclass of anything we might encode in our `NSAttributedString` attributes have had `+supportsSecureCoding` redefined

Indeed, apparently, **even if** the parent class already overrides it to return `true`, subclasses which override `init?(coder:)`/`encode(with:)` from their parent class need to also re-override `+supportsSecureCoding`. I've added the overrides in classes that I've modified, but there may be more classes that might not be covered by our unit tests around archiving/pasting but would still require it to be added? So would be worth making another pass to be sure we didn't forget any,

AliSoftware committed:
Use `UTType.utf8PlainText` (explicit encoding) instead of `kUTTypePlainText`

Since newer versions of iOS use this new `public.utf8-plain-text` UTI instead of `public.plain-text` like in previous OS versions.

AliSoftware committed:
Fix overriding non-open class outside module

AliSoftware committed:
Revert "Fix overriding non-open class outside module"

This reverts commit 5c3004b944eed56f286c7034ac02ea9d2716137c, because I'm not convinced that's a good idea.

We should double-check that **not** having `+supportsSecureCoding` declared on this subclass doesn't make it fail to be copy/pasted (i.e. archived/unarchived)—like similar cases happened when running Aztec tests. Or if we need `+supportsSecureCoding`, we need to find a way to override/redefine it across module boundaries…

AliSoftware committed:
Declare `class var supportsSecureCoding` as `open`

AliSoftware committed:
Simpler code style on UIPasteboard+Helpers > UTType

Commit:
Fix shorcode test by using an attribute with key-value

Commit:
Merge pull request #1395 from wordpress-mobile/fix/shortcode-test

Fix - Shortcode test

mokagio committed:
Add `shared-pipeline-vars` to Buildkite setup and use Xcode 16.2

mokagio committed:
Remove unnecessary `""` from Buildkite pipeline

This is in line with our Buildkite/YML pipeline standard.

mokagio committed:
Use inlined one-liner commands consistently in Buildkite pipeline

mokagio committed:
Use custom GitHub commit status labels for all steps

mokagio committed:
Run CI tests on iPhone 16 Pro iOS 18.2 Simulator

mokagio committed:
Explicitly set `device` in CI runs to avoid attempting macOS tests

mokagio committed:
Upgrade Fastlane to work with latest Apple toolchain

mokagio committed:
Upgrade to Xcode 16.2 (#1404)
